account_invoice_mode_weekly/models/sale_order.py

```diff
@@ -0,0 +1,76 @@
+# Copyright 2021 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html)
+
+from datetime import datetime
+
+from odoo import api, models
+
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.model
+    def cron_generate_weekly_invoices(self):
+        """Cron called daily to check if weekly invoicing needs to be done."""
+        company_ids = self._company_weekly_invoicing_today()
+        if company_ids:
+            self.generate_weekly_invoices(company_ids)
+
+    @api.model
+    def generate_weekly_invoices(self, companies=None):
+        """Generate weekly invoices for customers who require that mode.
+
+        Invoices will be generated by other jobs split for different customer
+        and different payment term.
+        """
+        if not companies:
+            companies = self.company_id
+        saleorder_groups = self.read_group(
+            [
+                ("invoicing_mode", "=", "weekly"),
+                ("invoice_status", "=", "to invoice"),
+                ("company_id", "in", companies.ids),
+            ],
+            ["partner_invoice_id"],
+            groupby=self._get_groupby_fields_for_weekly_invoicing(),
+            lazy=False,
+        )
+        for saleorder_group in saleorder_groups:
+            saleorder_ids = self.search(saleorder_group["__domain"]).ids
+            self.with_delay()._generate_invoices_by_partner(saleorder_ids)
+        companies.write({"invoicing_mode_weekly_last_execution": datetime.now()})
+        return saleorder_groups
+
+    @api.model
+    def _get_groupby_fields_for_weekly_invoicing(self):
+        """Returns the sale order fields used to group them into jobs."""
+        return ["partner_invoice_id", "payment_term_id"]
+
+    def _generate_invoices_by_partner(self, saleorder_ids, invoicing_mode="weekly"):
```

Comment:
@StefanRijnhart @jcoux After having read this, it's impossible to have several modes on the same instance as:

- This is should be implemented in base module
- This is always overridden without super in each specific module...

I'll try to fix that